server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeReplacementAllocationDecider.java

```diff
@@ -0,0 +1,94 @@
+            return Decision.single(Decision.Type.NO, NAME,
+                "node [%s] is replacing a vacating node, so no data from other nodes " +
+                    "may be allocated to it until the replacement is complete", node.nodeId());
+        } else {
+            return Decision.single(Decision.Type.YES, NAME,
+                "neither the source nor target node are part of an ongoing node replacement");
+        }
+    }
+
+    @Override
+    public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (replacementOngoing(allocation.metadata()) == false) {
+            return NO_REPLACEMENTS;
+        } else if (isReplacementSource(allocation.metadata(), node.nodeId())) {
+            return Decision.single(Decision.Type.NO, NAME,
+                "node [%s] is being replaced by node [%s], so no data may remain on it", node.nodeId(), "replacement");
+        } else {
+            return Decision.single(Decision.Type.YES, NAME, "node [%s] is not being replaced", node.nodeId());
+        }
+    }
+
+    /**
+     * See the comment in the else branch of {@link #canAllocate(ShardRouting, RoutingNode, RoutingAllocation)}
+     * for a reason why we allow allocation that may potentially allocate to the source of a node
+     * replacement shutdown.
+     */
+    @Override
+    public Decision canAllocate(IndexMetadata indexMetadata, RoutingNode node, RoutingAllocation allocation) {
+        if (replacementOngoing(allocation.metadata()) == false) {
+            return NO_REPLACEMENTS;
+        } else if (isReplacementTarget(allocation.metadata(), node.nodeId())) {
+            return Decision.single(Decision.Type.NO, NAME,
+                "node [%s] is replacing a vacating node, so no data from other nodes " +
+                    "may be allocated to it until the replacement is complete", node.nodeId());
+        } else {
+            // The node in question is not a replacement target, so allow allocation.
+            return Decision.single(Decision.Type.YES, NAME,
+                "node is not a replacement target, so allocation is allowed");
+        }
+    }
+
+    @Override
+    public Decision shouldAutoExpandToNode(IndexMetadata indexMetadata, DiscoveryNode node, RoutingAllocation allocation) {
+        if (replacementOngoing(allocation.metadata()) == false) {
+            return NO_REPLACEMENTS;
+        } else if (isReplacementTarget(allocation.metadata(), node.getId())) {
+            return Decision.single(Decision.Type.NO, NAME,
+                "node [%s] is a node replacement target, shards cannot auto expand to be on it until the replacement is complete",
+                node.getId(), "source");
+        } else if (isReplacementSource(allocation.metadata(), node.getId())) {
+            return Decision.single(Decision.Type.NO, NAME,
+                "node [%s] is being replaced, shards cannot auto expand to be on it", node.getId());
```

Comment:
I was on the fence about this also, I don't have a strong feeling about this, do you think we should go one way or another?

The difficulty with considering a source-target pair as one for auto-expand is that it could have been invalid to expand to the source node for other reasons, but valid on the target node, and that would but it in the "could be dangerous to add another shard to the target node" category. Hence, for safety I went with the "neither" approach for now.

Reply:
More than the classical "give me a copy on every node for search performance" I am more worried about our use of `0-1` on system indices like security indices.

In particular in the replace case, we risk that the reduce of replica count causes the remaining copy to reside on the source node, waiting to be vacated. Given that we are running against time here, that adds some risk of loosing the last copy of data. To become safe, we should also ensure that any copies on the target/source node is the one that we get rid of, which would be complicated.

Regarding your case of it being invalid on target node, I think there are really two cases there:

1. The shard already exists on the target node (i.e., the node was excluded somehow but the shard has not yet moved off).
2. The shard does not exist on the target node (but on some other nodes).

In neither case will we allow establishing a replica on the target node, `NodeReplacementAllocationDecider` will prevent that. Only relocations to target node will go through `canBalance`.

In the 1st case we would in many cases have to vacate the shard anyway. So the benefit is only in the case where we are lucky that the copy on the source node is the one we remove. Already an edge case since the shard is not supposed to reside on the target node altogether.

In the 2nd case there would be no additional benefit of not reducing replica count and no drawback of increasing it.

I suppose this is again a matter of data safety vs how sure we are the replace succeeds. I am inclined to prefer data safety :slightly_smiling_face:, in particular in this case since I think the drawbacks are minor and luck-related.